---
format:
    html:
        embed-rmd: true
        output-ext: "htm"
---

## Setting the stage
### DGE Instructions 
DGE comparison all clusters

```{r, eval=TRUE, echo=T, results=FALSE, message = FALSE, warning = FALSE}
###############################################################################
##  Load packages                                                            ##

version <- "dev"
a <- unlist(strsplit(getwd(), "/"))
folderNameTag <- a[length(a)]

VersionPdfExt <- paste0(folderNameTag,".",version,".V", gsub("-", "", Sys.Date()), ".pdf")

rasterDpi <- 300
calculateDGE <- TRUE

## Number of cells to use for downsamplng
nCellMax <- 20000




## Load bioLOGIC object
source("load.biologic.robj.R")

## Check if DGE_result table is available
if (!calculateDGE){
  if (is.null( Obio@dataTableList[["DGE_table"]])){
    calculateDGE <- TRUE
  }
}

## Make sure paths are set correctly in the Obio object
library(dplyr)
Obio <- Obio %>%
    biologicSeqTools2::setMountingPoint()  %>% 
    biologicSeqTools2::setAnalysisPaths() %>% 
    biologicSeqTools2::setCrickGenomeAndGeneNameTable() %>% 
    biologicSeqTools2::createAnalysisFolders() %>% 
    biologicSeqTools2::setDataBaseParameters()

## Load Seurat object ##
SeuratFN <- paste0(
    Obio@parameterList$localWorkDir,
    Obio@parameterList$project_id,
    ".Seurat.Robj"
)

load(SeuratFN)

## Convert to Seurat5 if it is not already##
# OsC <- Seurat::UpdateSeuratObject(object = OsC)

##                                                                           ##
###############################################################################
```
```{r, eval=TRUE, echo=T, results=FALSE, message = FALSE, warning = FALSE}
###############################################################################
## Subsetting Seurat object                                                  ##
# Sketching and Subsetting Seurat object

# options(Seurat.ojbect.assay.version = "v5")

## Object size
# format(Seurat::object.size(OsC), units = "Gb")

# format(object.size(OsC), units = "Gb")
# [1] "65 Gb"

## Sketch
pos <- grep("sketch", names(OsC@assays))

if (length(pos) == 0){
    OsC <- Seurat::SketchData(
        object = OsC,
        method = "LeverageScore",
        ncells = 20000,
        sketched.assay = "sketch"
    )
}
#
# ## Get cellIDs for sketch assay
Seurat::Idents(OsC) <- "sketch"
OsC <- SeuratObject::JoinLayers(object = OsC, layers = c("sketch", "RNA"))

sketchCellIDs <- colnames(OsC@assays$sketch)

if (length(sketchCellIDs) > nCellMax){
    sketchCellIDs <- sample(sketchCellIDs, nCellMax)
}



OsCsub <- subset(x = OsC, subset = cellID %in% sketchCellIDs)
rm(OsC)
OsC <- OsCsub

Seurat::Idents(OsC) <- "RNA"

#
# ## Object size
format(object.size(OsCsub), units = "Gb")
# # [1] "29.3 Gb"

##                                                                           ##
###############################################################################

```

```{r, eval=TRUE, echo=T, results=FALSE, message = FALSE, warning = FALSE}
###############################################################################
## Display cluster Names                                                     ##

dfDis <- unique(OsC@meta.data[,c("clusterName", "seurat_clusters")])

dfDis$clusterName <- factor(dfDis$clusterName, levels = Obio@parameterList$clusterNameOrder)
rownames(dfDis) <- NULL
dfDis <- dfDis[order(dfDis$clusterName),]

DT::datatable(dfDis)
```

```{r, eval=TRUE, echo=T, results=FALSE, message = FALSE, warning = FALSE}

###############################################################################
## Retrieve database password                                                ##
if (dir.exists("/Volumes/babs/working/boeings/")){
    hpc.mount <- "/Volumes/babs/working/boeings/"
} else if (dir.exists("Y:/working/boeings/")){
    hpc.mount <- "Y:/working/boeings/"
} else if (dir.exists("/camp/stp/babs/working/boeings/")){
    hpc.mount <- "/camp/stp/babs/working/boeings/"
} else {
    hpc.mount <- ""
}

FN <- paste0(hpc.mount, "Projects/reference_data/pwd_folder/babs.txt")
    dbTable <- read.delim(
      FN,
      header = F,
      sep = "\t",
      stringsAsFactors = F
    )
#}
db.pwd <- as.vector(dbTable[1,1])

## Done                                                                      ##
###############################################################################

###############################################################################
##  Set variables                                                            ##

# Plotting parameters
figureCount <- 1
legendDotSize <- 5

## Define project ID ##
project_id <- Obio@parameterList$project_id


## Set directory for report tables
reportTableDir <- Obio@parameterList$reportTableDir
reportFigDir <- Obio@parameterList$reportFigDir

localWorkDir <- Obio@parameterList$localWorkDir

## Create url string
if (Obio@parameterList$host == "10.27.241.234"){
    urlString <- "biologic.thecrick.org"
} else {
    urlString <- "biologic.crick.ac.uk"
}

shinyBaseUrl <- "https://bioinformatics.crick.ac.uk/shiny/users/boeings/"

shinyURL <- paste0(
    shinyBaseUrl,
    project_id,
    "_app/"
)            
   
## Set file paths ##
baseFN <- paste0(
   project_id, 
   ".", folderNameTag,".table.",
   version,
   ".xlsx"
)

FNrel <- paste0("report_tables/", baseFN)

outPutFN <- paste0(
     reportTableDir,
     baseFN
)

tableLink <- paste0(
    'https://',
    urlString,
    '/mdata/',project_id, '/html/', 
    FNrel
)



tableString <- paste0(
    'An Excel table with the DGE results can be downloaded ',
    tableLink
)

tableString <- paste0(
   ' An Excel table with the DGE results can be downloaded <a href="',tableLink,'" target="_blank">here</a>. '
)
  

##  Set variables                                                            ##
###############################################################################



###############################################################################
## Function doDGEsc                                                          ##

#' @title doDGEsc
#' @description Function to do glmgampoi differential gene expression. When using this function, make sure the glmGamPoi R-package is installed on your system: renv::install("renv::install("bioc::glmGamPoi") 
#' @param obj Seurat object for the DGE analysis
#' @param countSlot Seurat object assay to use to retrieve count data.
#' @param DGEselCol Meta data column in the above Seurat object to use for DGE sample group definitions. 
#' @param colName Name for the DGE comparison in the meta data
#' @param contrastTag contrastTag
#' @param DGEsampleList List specifying the DGE comparisons. 
#' @import Seurat
#' @export


setGeneric(
    name="doDGEsc",
    def=function(
        obj,
        countSlot = "RNA",
        DGEselCol = "sub_clusters_ExNeurons",
        colName = "DGE_name",
        contrastTag = "contrast_1_",
        DGEsampleList = list(
            "M1" = c(5),
            "M2" = c(1)
        )
    ) {
    
    ############################################################################
    ## Add DGE cell selection to meta-data                                    ##
    colName <- unlist(colName)
    DGEselCol <- unlist(DGEselCol)

    pos <- grep(colName, names(obj@meta.data))
    if (length(pos) > 0){
        obj@meta.data <- obj@meta.data[,-pos]
    }

    mColName <- substr(paste0("meta_", colName), 1, 63)
    obj@meta.data[[mColName]] <- "Rest" 
    
    for (l in 1:length(DGEsampleList)){
        obj@meta.data[obj@meta.data[,DGEselCol] %in% DGEsampleList[[l]], mColName] <- names(DGEsampleList)[l]
    }
    
    dfMeta <- obj@meta.data
    dfMeta <- dfMeta[dfMeta[,mColName] != "Rest",]
    
    ##                                                                        ##
    ############################################################################
    
    ############################################################################
    ## Get counts for DGE                                                     ##
    cellVec <- dfMeta$cellID
    group <-  as.vector(dfMeta[,mColName])
    
    dfMatrix <- OsC@assays$RNA$counts
    dfMatrix <- data.matrix(dfMatrix[,cellVec])
    dfMatrix <- dfMatrix[rowSums(dfMatrix) != 0, ]
    
    ## Pseudobulk DGE
    
    ## LRT ##
    # fit <- glm_gp(dfMatrix, design = group)
    # res <- test_de(fit, reduced_design = ~ 1)
    
    sample_labels <- group
    
    sample_labels[sample_labels == names(DGEsampleList)[1]] <- paste0(sample_labels[sample_labels == names(DGEsampleList[1])], "_", 1:length(sample_labels[sample_labels == names(DGEsampleList)[1]]))
    
    sample_labels[sample_labels == names(DGEsampleList)[2]] <- paste0(sample_labels[sample_labels == names(DGEsampleList[2])], "_", 1:length(sample_labels[sample_labels == names(DGEsampleList)[2]]))
    
    cell_type_lables <- group
    unique(group)
    
    fit <- glmGamPoi::glm_gp(
        dfMatrix, 
        design = group
    )
    
    comparison <- names(DGEsampleList)
    
    contrastString <- (paste0(comparison[1], " - ", comparison[2]))
    
    res <- glmGamPoi::test_de(fit, contrast = contrastString,
            pseudobulk_by = sample_labels, 
            #subset_to = cell_type_labels == "T-cells",
            #n_max = 4, 
            sort_by = pval, 
            decreasing = FALSE
    )
    
    dfRes <- res[order(res$lfc),]
    
    ## Remove extreme outliers ##
    dfRes <- dfRes[dfRes$lfc > -100 & dfRes$lfc < 100, ]
    dfRes[["padj"]] <- dfRes$adj_pval
    minP <- min(dfRes$padj[dfRes$padj != 0])
    dfRes[dfRes$padj == 0, "padj"] <- minP
    dfRes[["lg10p"]] <- -1 * log10(dfRes$padj)
    
    
    comp_1 <- dfRes
    comp_1[["gene"]] <- dfRes$name
    names(comp_1) <- gsub("lfc", paste0(contrastTag, "logFC_" ,colName), names(comp_1))
    
    names(comp_1) <- gsub("padj", paste0(contrastTag, "padj_",colName), names(comp_1))
    names(comp_1) <- gsub("lg10p", paste0(contrastTag, "lg10p_",colName), names(comp_1))
    
    selVec <- c(
        "gene",
        names(comp_1)[grep(contrastTag, names(comp_1))]
    )
    
    comp_1 <- unique(comp_1[,selVec])
    
    ## Cut all strings in comp_1 to a maximum of 63 charcters 
    names(comp_1) <- sapply(names(comp_1), function(x) substr(x, 1, 63))
    
    
    
    ##                                                                        ##
    ############################################################################
    
    
    
    returnList <- list(
      "OsC" = obj,
      DGEtable = comp_1
    )
    
    return(returnList)
})


## End doDGEsc Function                                                      ##
###############################################################################

###############################################################################
## Define differential gene expression comparisons in a list                 ##
## One list entry per comparison                                             ##

## Create custom column for this project:
# samplegroup_seurat_clusters

###############################################################################
## Define differential gene expression comparisons in a list                 ##
## One list entry per comparison                                             ##

## Create custom column for this project:
# samplegroup_seurat_clusters
compCol1 <- "clusterName"
compCol2 <- "sampleName"
segmentationCol <- paste0(compCol1, "_", compCol2)

OsC@meta.data[[segmentationCol]] <- paste0(
    OsC@meta.data[[compCol1]],
    "_",
    OsC@meta.data[[compCol2]]
)



dfTemp <- unique(OsC@meta.data[,c(compCol1, compCol2, segmentationCol)])

dfTemp <- dfTemp[order(dfTemp[,compCol1]),]

## Optional
#selVec <- c(0, 2, 8, 10)
## 8 and 10 only exist in tumor
#selVec <- c(0, 2, 3)
#dfTemp <- dfTemp[dfTemp$seurat_clusters %in% selVec, ]

comparisonCol <- "clusterName"
groupCol <- segmentationCol

DGEoptions <- unique(dfTemp[,comparisonCol])

compList <- list()

## Make DGE list
for (i in 1:length(DGEoptions)){

    tag <- paste0("comp_", i)

    dfTemp2 <- dfTemp[grep(as.vector(paste0("^", DGEoptions[i], "$")), dfTemp[,comparisonCol]),]

    if (nrow(dfTemp2) == 2){
      groupA <- dfTemp2[2, groupCol ]
      groupB <- dfTemp2[1, groupCol ]
      name <- paste0("DGE_", groupA, "_vs_", groupB)

      Ldge <- list(
          "pert" = c(groupA),
          "baseline" = c(groupB)
      )

      names(Ldge) <- c(groupA, groupB)

      compList[[tag]] <- list(
          "name"    = name,
          "DGEselCol" = groupCol,
          "DGEsampleList" =  Ldge

      )

      names(compList)


    } else {
      stop("Reconfiguration L351 required")
    }


    
}


DGEtagVec <- as.vector(NULL, mode = "character")


DGEinputList <- compList


###############################################################################
## Function do scDGE                                                         ##

doScDGEonList <- function(
    OsC,
    tag,
    DGEinputList,
    DGEtagVec = as.vector(NULL, mode = "character"),
    i
    
){
    resList <- doDGEsc(
        obj = OsC,
        DGEselCol = unlist(DGEinputList[[i]]["DGEselCol"]),
        colName = unlist(DGEinputList[[i]]["name"]),
        contrastTag = paste0("contrast_",i,"_"),
        DGEsampleList = DGEinputList[[i]][["DGEsampleList"]]
    )
    
    OsC <- resList$OsC
    
    ###########################################################################
    ## Add average intensity for MA-Plot to DGE table                        ##
    
    OsC_DGE <- OsC
    OsC_DGE@meta.data[["DGE_sel"]] <- "P0"
    
    OsC_DGE@meta.data[OsC_DGE@meta.data[,unlist(DGEinputList[[i]]["DGEselCol"])] != "", "DGE_sel"] <- "P1"
    
    #row.names(OsC_DGE@meta.data) <- OsC_DGE@meta.data[,"cellID"]
    
    OsC_DGE <- subset(x = OsC_DGE, subset = DGE_sel == "P1")
    
    


   ## Seurat5 at present can't calculate average expression if only one value is given
   ## workaround ##
   if (length(unique(OsC_DGE@meta.data[["DGE_sel"]])) == 1){
        OsC_DGE@meta.data[1, "DGE_sel"] <- "Added"
   }
   Seurat::Idents(OsC_DGE) <- "DGE_sel"

    cluster.averages <- Seurat::AggregateExpression(
        OsC_DGE, 
        return.seurat = TRUE
    )

    if (packageVersion("Seurat") > "4.8.0"){
        dfAvgExpr <- data.frame(cluster.averages@assays$RNA$data)
    } else {
        dfAvgExpr <- data.frame(cluster.averages[["RNA"]]@data)
    }

    dfAvgExpr[["gene"]] <- row.names(dfAvgExpr)
    dfAvgExpr$Added <- NULL
    


    row.names(dfAvgExpr) <- NULL
    
    nameTag <- paste0("add_MA_cts_MA_Avg_", unlist(DGEinputList[[i]]$name))
    names(dfAvgExpr) <- gsub("P1", nameTag, names(dfAvgExpr))
    
    dfDataTable <- resList$DGEtable
    
    dfDataTable <- merge(
        dfDataTable, 
        dfAvgExpr, 
        by.x = "gene",
        by.y = "gene",
        all =TRUE
    )
    
    dfDataTable[is.na(dfDataTable)] <- 0
    names(dfDataTable) <- sapply(names(dfDataTable), function(x) substr(x, 1, 63))
    
    DGEtagVec <- c(
      DGEtagVec,
      tag
    )
    
    resultList <- list(
        "dfDataTable" = dfDataTable,
        "DGEtagVec" = DGEtagVec,
        "OsC" = OsC
    )
    
    return(resultList)
    
}

## Done do scDGE                                                             ##
###############################################################################

###############################################################################
## DGE Analysis                                                              ##

DGEtagVec = as.vector(NULL, mode = "character")
plotList <- list()
chnkVec <- as.vector(NULL, mode = "character")

## This does the DGE and creates an overview UMAP figure ##

## Downsize to 100k cells if the Seurat object is very large. 


for (k in 1:length(DGEinputList)){
    
    tag <- paste0(unlist(DGEinputList[[k]]["name"]))
    
    if (calculateDGE){
      resList <- doScDGEonList(
          OsC = OsC,
          tag = tag,
          DGEinputList = DGEinputList,
          DGEtagVec = DGEtagVec,
          i = k
      )
    
      dfDataTable <- resList[["dfDataTable"]]
      DGEtagVec <- resList[["DGEtagVec"]]
      OsCsub <- resList[["OsC"]]
      
      ## Add results to biologic object 
      Obio@dataTableList[[tag]] <- data.frame(NULL)
      Obio@dataTableList[[tag]] <- dfDataTable
      
      ## Done with DGE Analysis                                                ##
      ###########################################################################
    } else {
      DGEtagVec <- c(
        DGEtagVec, 
        tag
      )
    }
    
    ###########################################################################
    ## Create DGE Selection plot                                             ##
    dfPlot <- OsCsub@meta.data
    
    
    pos <- grep("included", names(dfPlot))
    if (length(pos) == 0){
        dfPlot[["included"]] <- "+"
    }

    pos <- grep("cellID", names(dfPlot))
      if (length(pos) == 0){
      dfPlot[["cellID"]] <- row.names(dfPlot)
    }

    dfPlot$UMAP_1 <- NULL
    dfPlot$UMAP_2 <- NULL
            
    ## Get UMAP coordinates ##
    coord <- data.frame(OsCsub@reductions$umap@cell.embeddings)
    coord[["cellID"]] <- row.names(coord)
    coord <-coord[coord$cellID %in% dfPlot$cellID, ]
            
    dfPlot <- merge(dfPlot, coord, by.x = "cellID", by.y="cellID", all=T)
    dfPlot[is.na(dfPlot)] <- 0
    dfPlot <- dfPlot[dfPlot$UMAP_1 != 0 & dfPlot$UMAP_2 != 0,]
            
            
    ## Add cluster colors ##
    dfPlot[["Cluster"]] <- dfPlot[,paste0("meta_", unlist(DGEinputList[[k]]["name"]))]
    dfPlot$Cluster[is.na(dfPlot$Cluster)] <- ""
    dfPlot[dfPlot$Cluster == "", "Cluster"] <- "Rest"
    
    label <- as.vector(gsub("DGE_", "", unlist(DGEinputList[[i]]["name"])))
    
    
    redGroup <- names(DGEinputList[[k]]$DGEsampleList)[1] #as.vector(unlist(strsplit(label, "_vs_"))[1])
    blueGroup <- names(DGEinputList[[k]]$DGEsampleList)[2] #as.vector(unlist(strsplit(label, "_vs_"))[2])
    greyGroup <- "Rest"
    
    ## Assigning colors ##
    dfCol <- data.frame(name = unique(dfPlot$Cluster), col = rep("", length(unique(dfPlot$Cluster))), stringsAsFactors = F)
    dfCol[dfCol$name == redGroup, "col"] <- "#000080"
    dfCol[dfCol$name == blueGroup, "col"] <- "#009900"
    dfCol[dfCol$name == greyGroup, "col"] <- "grey"
    colVec <- dfCol$col
    names(colVec) <- dfCol$name
    
    dfColExtra <- dfCol[dfCol$col == "",]
    if (nrow(dfColExtra) > 0){
        eColVec <- dfColExtra$name
        library(scales)
        eCols = hue_pal()(length(eColVec))
        for (l in 1:length(eColVec)){
            dfCol[dfCol$name == eColVec[l], "col"] <- eCols[l]
        }
    }
    
    maxX <- 1.1*max(dfPlot$UMAP_1, na.rm = T)
    minX <- 1.1*min(dfPlot$UMAP_1, na.rm = T)
    maxY <- 1.1*max(dfPlot$UMAP_2, na.rm = T)
    minY <- 1.1*min(dfPlot$UMAP_2, na.rm = T)            

    #colVec <- dfPair$sampleColor
    #names(colVec) <- dfPair$sampleName
    levels <- as.vector(dfCol$name)
    levelsRest <- levels[grep("Rest", levels)]
    levels <- levels[levels != "Rest"]
    levels <- c(levels, levelsRest)
    
    dfPlot$Cluster <- factor(dfPlot$Cluster, levels = levels)

    dotsize  = 1
    if (nrow(dfPlot) > 10000){
        dotsize  = 0.5
    } else if (nrow(dfPlot) > 20000){
        dotsize = 0.25
    } else if (nrow(dfPlot) > 50000){
        dotsize = 0.1
    }

    ## make plot ##
    plotList[[tag]] <- ggplot2::ggplot(
        data=dfPlot[dfPlot$included == "+",], 
        ggplot2::aes(UMAP_1, UMAP_2, color=Cluster)) + 
        ggrastr::rasterize(ggplot2::geom_point( shape=16, size = as.numeric(dotsize)), dpi = rasterDpi) + 
        ggplot2::xlab("UMAP1") + 
        ggplot2::ylab("UMAP2") + 
        ggplot2::theme_bw()  +  
        ggplot2::theme(
          axis.text.y   = ggplot2::element_text(size=8),
          axis.text.x   = ggplot2::element_text(size=8),
          axis.title.y  = ggplot2::element_text(size=8),
          axis.title.x  = ggplot2::element_text(size=8),
          axis.line = ggplot2::element_line(colour = "black"),
          panel.border = ggplot2::element_rect(colour = "black", fill=NA, size=1),
          plot.title = ggplot2::element_text(hjust = 0.5, size = 12),
          legend.title = ggplot2::element_blank()
        ) + 
        ggplot2::guides(col = ggplot2::guide_legend(override.aes = list(shape = 16, size = legendDotSize))) +     
        ggplot2::ggtitle(paste0("Cell Selection for ", gsub("_", " ", unlist(DGEinputList[[k]]["name"])))) +
        ggplot2::xlim(minX, maxX) + 
        ggplot2::ylim(minY, maxY) + 
        ggplot2::coord_fixed(ratio=1) + 
        ggplot2::scale_colour_manual("DGE Groups" ,values = colVec) + 
        ggplot2::guides(col = ggplot2::guide_legend(override.aes = list(shape = 16, size = legendDotSize)))


      ## Add colors if specified ##

            
      ## Make plot figure ##
      FNbase <- paste0(tag, VersionPdfExt)
      FN <- paste0(reportFigDir, FNbase)
      FNrel <- paste0("report_figures/", FNbase)
      
      ## Print high quality pdf        
      pdf(FN)
          print(plotList[[tag]])
      dev.off()
      
      ## downsize for report plot
      # plotList[[tag]] <- ggrastr::rasterize(
      #   plotList[[tag]],
      #   layers='Point',
      #   dpi = 700
      # )
              
        # link <- paste0('<a href="https://',urlString,'/',Obio@parameterList$project_id,'/pca?x_axis=UMAP_1&y_axis=UMAP_2" target="_blank">here</a>')  
        
      if (exists("shinyURL") & !is.null(shinyURL)){
          link <- paste0(
                  'An interactive version of this figure with additional viewing options can be found <a href="',shinyURL,'?_inputs_&y_axis=%22UMAP_2%22&x_axis=%22UMAP_1%22&colorBy=%22',paste0("meta_", tag),'%22&splitByColumn=%22all%22" target="_blank">here</a>. '
          )
                
      } else {
          link <- ""
      }
              
      figLegend <- paste0(
      '**Figure ', 
      figureCount, 
                ':** ',
                ' UMAP showing all cells from all samples together. Download a pdf of this figure <a href="',FNrel,'" target="_blank">here</a>. ',
                'An interactive version of this figure can be found ', link
            )
            
      figureCount <- figureCount + 1
            
      NewChnk <- paste0(
          "#### ", tag,
          "\n```{r Sample_DGE_",
          tag,", results='asis', echo=F, eval=TRUE, warning=FALSE, fig.cap='",
          figLegend,"'}\n",
          "\n",
          "\n print(plotList[['",tag,"']])",
          "\n cat(  '\n')",
          "\n\n\n```\n"   
      )
            
      chnkVec <- c(
          chnkVec,
          NewChnk
      )
  
      ## Done creating DGE selection plot                                      ##
      ###########################################################################
      
      
      print(paste0("Done with ", tag))
}

## Done DGE                                                                  ##
###############################################################################

###############################################################################
## Add DGE results to object                                                 ##

###############################################################################
## Assemble DGE summary table                                                ##
#dfTemp <- Obio@dataTableList[["DGE_table"]]

wb <- openxlsx::createWorkbook()

hs1 <- openxlsx::createStyle(
    fontColour = "#ffffff",
    fgFill = "#000000",
    halign = "CENTER",
    textDecoration = "Bold"
)

###############################################################################
## Add Create DGE result Excel Spreadsheet                                   ##
## Add additional filters



for (i in 1:length(DGEtagVec)){

  ## Add to Excel workbook ##
  dfOutput <-   Obio@dataTableList[[DGEtagVec[i]]]
  orderVec <- names(dfOutput)
  dfOutput[["gene_FeatureViewLink"]] <- paste0(
      shinyURL,
      "?gene=",
      dfOutput$gene
  )
  orderVec <- orderVec[orderVec != "gene"]
  orderVec <- c("gene", "gene_FeatureViewLink", orderVec)
  dfOutput <- dfOutput[, orderVec]
  class(dfOutput$gene_FeatureViewLink) <- c(class(dfOutput$gene_FeatureViewLink), "hyperlink")

  ## Highlight logFCtwoSD
  logFCtwoSD <- 2*sd(dfOutput[,grep("_logFC_", names(dfOutput))])
  dfOutput[["DGE_Status"]] <- "Unchanged"
  dfOutput[dfOutput[,grep("_logFC_", names(dfOutput))] > logFCtwoSD & dfOutput[,grep("_padj_", names(dfOutput))] < 0.05, "DGE_Status"] <- "Up"
    dfOutput[dfOutput[,grep("_logFC_", names(dfOutput))] < -1 *logFCtwoSD & dfOutput[,grep("_padj_", names(dfOutput))] < 0.05, "DGE_Status"] <- "Down"

  if (nrow(dfOutput[dfOutput$DGE_Status != "Unchanged",]) > 0){
    dfOutput <- dfOutput[dfOutput$DGE_Status != "Unchanged",]
  }

  ## Highlight TFs
  if (!is.null(Obio@dataTableList[["referenceList"]][["TFs"]])){
    geneSet <- Obio@dataTableList[["referenceList"]][["TFs"]]
    dfOutput[["Transcription_factor"]] <- ""
    dfOutput[dfOutput$gene %in% geneSet, "Transcription_factor"] <- "+"
  }



  ## Sort based on logFCs
  pos <- grep("logFC_", names(dfOutput))
  if (length(pos) == 1){
    dfOutput <- dfOutput[order(dfOutput[, pos], decreasing = T),]
  }

  ## Make sure sheetname is unique
  sheetName <- paste0(
    substr(
      DGEtagVec[i],1,27
    ),
    "_", i
  )

  openxlsx::addWorksheet(
      wb,
      sheetName = sheetName
  )


  openxlsx::freezePane(wb, sheetName ,  firstActiveRow = 2)
  openxlsx::writeData(wb, sheetName, dfOutput, startRow = 1, startCol = 1, headerStyle = hs1)
  openxlsx::addFilter(wb, sheetName, row = 1, cols = 1:ncol(dfOutput))
  ## Done adding to Excel workbook ##

  if (i ==1){
      dfTemp <- Obio@dataTableList[[DGEtagVec[i]]]
  } else {
      dfTemp <- merge(
        dfTemp,
        Obio@dataTableList[[DGEtagVec[i]]],
        by.x = "gene",
        by.y = "gene",
        all =TRUE
      )
      dfTemp[is.na(dfTemp)] <- 0

  }

  ## Create long-format marker gene list
  dfOutput[["cluster"]] <- DGEtagVec[i]
  names(dfOutput) <- c(
    "gene",
    "gene_FeatureViewLink",
    "logFC",
    "padj",
    "lg10p",
    "avg_intensity",
    "DGE_Status",
    "Transcription_factor",
    "cluster"
  )

  dfOutput[["intensity_logFC"]] <- dfOutput$avg_intensity * dfOutput$logFC
  dfOutput <- dfOutput[order(dfOutput$intensity_logFC, decreasing = T), ]
  dfOutput[["markerRank"]] <- 1:nrow(dfOutput)
  dfOutput[["clusterName"]] <- dfOutput$cluster
  dfOutput$clusterName <- gsub("DGE_", "", dfOutput$clusterName)
  dfOutput$clusterName <- gsub("_vs_other_clusters", "", dfOutput$clusterName)


  if (i ==1){
    dfLongMarker <- dfOutput
  } else {
    dfLongMarker <- rbind(
      dfLongMarker,
      dfOutput
    )
  }
}

## Save workbook ##
baseFN <- paste0(
   project_id,
   ".",folderNameTag,".table.",
   version,
   ".xlsx"
)


outPutFN <- paste0(
     reportTableDir,
     baseFN
)

if (!dir.exists(reportTableDir)){
    dir.create(reportTableDir, recursive=T)
}

openxlsx::saveWorkbook(
        wb,
        outPutFN ,
        overwrite = TRUE
)

############################################3
## Save long format workbook
wb <- openxlsx::createWorkbook()

hs1 <- openxlsx::createStyle(
    fontColour = "#ffffff",
    fgFill = "#000000",
    halign = "CENTER",
    textDecoration = "Bold"
)

sheetName = "ClusterMarkers"
  openxlsx::addWorksheet(
      wb,
      sheetName = sheetName
)


openxlsx::freezePane(wb, sheetName ,  firstActiveRow = 2)
openxlsx::writeData(wb, sheetName, dfLongMarker, startRow = 1, startCol = 1, headerStyle = hs1)
openxlsx::addFilter(wb, sheetName, row = 1, cols = 1:ncol(dfOutput))

## Save workbook ##
baseFN <- paste0(
   project_id,
   ".",folderNameTag,".long.marker.table.",
   version,
   ".xlsx"
)


outPutFN <- paste0(
     reportTableDir,
     baseFN
)

if (!dir.exists(reportTableDir)){
    dir.create(reportTableDir, recursive=T)
}

openxlsx::saveWorkbook(
        wb,
        outPutFN ,
        overwrite = TRUE
)







## Done creating xlsx spreadsheet                                            ##
###############################################################################

## Add to DGE table
if (!is.null(Obio@dataTableList[["DGE_table"]]) & is.data.frame(Obio@dataTableList[["DGE_table"]])) {
  dfDGE_old <- Obio@dataTableList[["DGE_table"]]
  colVec <- names(dfTemp)
  colVec <- colVec[colVec != "gene"]

  ## remove all columns with the same name, keep the rest
  selCols <- names(dfDGE_old)[!(names(dfDGE_old) %in% colVec)]

  if (length(selCols) > 1){
    dfDGE_old <- dfDGE_old[,selCols]
    dfDGE_new <- merge(
      dfDGE_old,
      dfTemp,
      by.x = "gene",
      by.y = "gene",
      all = T
    )
  } else {
    dfDGE_new <- dfTemp
  }

} else {
    dfDGE_new <- dfTemp
}

Obio@dataTableList[["DGE_table"]] <- NULL
Obio@dataTableList[["DGE_table"]] <- dfDGE_new

## save changes ##
if (calculateDGE){
  source("save.biologic.robj.R")
}

# dfDGEstore <- dfTemp

## Done                                                                      ##
###############################################################################



if (length(plotList) > 3){
    tabVar <- ".tabset .tabset-fade .tabset-dropdown"
} else {
    tabVar <- ".tabset .tabset-fade .tabset-pills"
}
``` 

### Differential Gene Expression Comparions Cell Selections {`r tabVar`}
In this section the cells that were used for the DGE comparisons are highlighted.

:::: {.panel-tabset .nav-pills}

```{r CellTypePlot, echo=TRUE, eval=TRUE, warning=FALSE, results="asis"}
cat(paste(knitr::knit(text = chnkVec, quiet = T), collapse = '\n'))
``` 
::::

```{r, eval=TRUE, echo=F, results=FALSE, message = FALSE, warning = FALSE}
tempDir <- "../../../../workdir/temp/"

if (!dir.exists(tempDir)){
  dir.create(tempDir, recursive = T)
}

FN <- "../../../../workdir/temp/temp.workspace.RData"
save.image(FN)

```

